### PR TITLE
http: add Kill Request HTTP filter

### PR DESCRIPTION
**Pull request by qqustc**
Add a KillRequest HTTP filter which can crash Envoy when receiving a Kill request. It will be used to fault inject kill request to Envoy and measure the blast radius.
The new KillRequest filter will be disabled at build time by default in Envoy by comment out the KillRequest filter extension in extensions_build_config.bzl

Risk Level: Low, new feature.
Testing: Unit/integration tests.
Docs Changes: Added
Release Notes:
Issue: #13978

### CODE REVIEW

**Comment by qqustc** [comment]
\retest

File: test/extensions/filters/http/kill_request/kill_request_config_test.cc [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  Http::FilterFactoryCb cb =
+      factory.createFilterFactoryFromProto(*factory.createEmptyConfigProto(), "stats", context);
+  Http::MockFilterChainFactoryCallbacks filter_callback;
+  EXPECT_CALL(filter_callback, addStreamFilter(_));
+  cb(filter_callback);
+}
+
+// Test that the deprecated extension name still functions.
+TEST(KillRequestConfigTest, DEPRECATED_FEATURE_TEST(DeprecatedExtensionFilterName)) {
+  const std::string deprecated_name = "envoy.kill_request";
```
Comment (htuch) [inline test/extensions/filters/http/kill_request/kill_request_config_test.cc:44]: You shouldn't need to set a deprecated name for a new extension.

File: source/extensions/filters/http/kill_request/kill_request_config.cc [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    callbacks.addStreamFilter(
+        std::make_shared<KillRequestFilter>(proto_config, context.api().randomGenerator()));
+  };
+}
+
+/**
+ * Static registration for the KillRequest filter. @see RegisterFactory.
+ */
+REGISTER_FACTORY(KillRequestFilterFactory,
+                 Server::Configuration::NamedHttpFilterConfigFactory){"envoy.kill_request"};
```
Comment (htuch) [inline source/extensions/filters/http/kill_request/kill_request_config.cc:27]: See comment below about deprecated names. In fact, all extension names are now effectively deprecated. @kyessenov should we have some way to register without declaring these at all?

File: source/extensions/filters/http/kill_request/kill_request_filter.cc [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  uint32_t denominator = 100;
+  if (kill_request_.probability().denominator() ==
+      envoy::type::v3::FractionalPercent::TEN_THOUSAND) {
+    denominator = 10000;
+  } else if (kill_request_.probability().denominator() ==
+             envoy::type::v3::FractionalPercent::MILLION) {
+    denominator = 1000000;
+  }
+
+  return (random_generator_.random() % denominator) < kill_request_.probability().numerator();
```
Comment (htuch) [inline source/extensions/filters/http/kill_request/kill_request_filter.cc:22]: See existing utilities for this, `ProtobufPercentHelper::evaluateFractionalPercent` etc.

**Review by htuch — CHANGES_REQUESTED** [review]
Thanks. I'll assume there is no strong objection around general utility and review. I think we should have some CI test to verify this is not included in any normal Envoy binary, or some Bazel magic to make sure it only comes up with a specific `--define`.
/wait

File: docs/root/configuration/http/http_filters/kill_request_filter.rst [hunk 4]
```diff
@@ -0,0 +1,39 @@
+.. _config_http_filters_kill_request:
+
+Kill Request
+===============
+
+The KillRequest filter can be used to crash Envoy when receiving a Kill request
+and measure the blast radius. The KillRequest filter is disabled at build time by default.
```
Comment (oschaaf) [inline docs/root/configuration/http/http_filters/kill_request_filter.rst:7]: drive by comment: what does blast radius mean in this context?

File: source/extensions/filters/http/kill_request/kill_request_config.cc (hunk 2 above)
Comment (qqustc) [reply source/extensions/filters/http/kill_request/kill_request_config.cc:27]: Yes, we can just do `REGISTER_FACTORY(KillRequestFilterFactory, Server::Configuration::NamedHttpFilterConfigFactory)`
https://github.com/envoyproxy/envoy/blob/9ccd778d4d7ccd986419620ed2e93b9796116863/include/envoy/registry/registry.h#L504

File: test/extensions/filters/http/kill_request/kill_request_config_test.cc (hunk 1 above)
Comment (qqustc) [reply test/extensions/filters/http/kill_request/kill_request_config_test.cc]: Removed this test.

File: source/extensions/filters/http/kill_request/kill_request_filter.cc (hunk 3 above)
Comment (qqustc) [reply source/extensions/filters/http/kill_request/kill_request_filter.cc]: Thanks Harvey, switched to evaluateFractionalPercent

File: docs/root/configuration/http/http_filters/kill_request_filter.rst (hunk 4 above)
Comment (qqustc) [reply docs/root/configuration/http/http_filters/kill_request_filter.rst:7]: Thanks Otto. Here the blast radius means when we send a kill request to 1 Envoy among a pool of Envoys, how many neighboring Envoy will be affected. This is specific to the downstream use case, so will remove this from the doc.

File: source/extensions/filters/http/kill_request/kill_request_config.cc (hunk 2 above)
Comment (kyessenov) [reply source/extensions/filters/http/kill_request/kill_request_config.cc]: Yeah, the name argument here is supposed to be an alias I think. Factories have unique names with aliases.

**Comment by qqustc** [comment]
> Thanks. I'll assume there is no strong objection around general utility and review. I think we should have some CI test to verify this is not included in any normal Envoy binary, or some Bazel magic to make sure it only comes up with a specific `--define`.
> /wait

Thank you Harvey! I have commented out the new filter in extensions_build_config.bzl and added a test in `main_common_test.cc` to verify the filter is not built into Envoy by default.

**Comment by qqustc** [comment]
> > Thanks. I'll assume there is no strong objection around general utility and review. I think we should have some CI test to verify this is not included in any normal Envoy binary, or some Bazel magic to make sure it only comes up with a specific `--define`.
> > /wait
> 
> Thank you Harvey! I have commented out the new filter in extensions_build_config.bzl and added a test in `main_common_test.cc` to verify the filter is not built into Envoy by default.

It turns out CI Generate docs failed after we commented out the new filter in `extensions_build_config.bzl` because [it can't generate the rst doc for `kill_request.proto`](https://dev.azure.com/cncf/envoy/_build/results?buildId=58458&view=logs&j=91e633a5-4907-5da1-5862-92a79fe2387a&t=f29e24e7-a093-5724-80bd-d79975d268cb)
So instead we modify the definition of `envoy_all_extensions()` and `envoy_all_http_filters()` in `all_extensions.bzl` which is equivalent to commenting out the new filter in `extensions_build_config.bzl`.

**Comment by qqustc** [comment]
Sorry for the spam, accidentally messed up this PR, closed this PR and will open a new one

**Pull request closed without merging**